From 98d80fce1086fc5c3d38e7f8a075c5fc5466d8c4 Mon Sep 17 00:00:00 2001
From: Brian Ryner <bryner@google.com>
Date: Mon, 26 Apr 2021 10:36:55 +1000
Subject: [PATCH 5/6] Fix test.support.fd_count() on macOS 11.

Trying to dup() all possible FDs causes the process to be killed by the
system. Instead, use the code path that counts file descriptors under /dev/fd.
---
 Lib/test/support/__init__.py | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/Lib/test/support/__init__.py b/Lib/test/support/__init__.py
index ccc11c1b4b..ffc252a6f6 100644
--- a/Lib/test/support/__init__.py
+++ b/Lib/test/support/__init__.py
@@ -2081,11 +2081,12 @@ def _crash_python():
 def fd_count():
     """Count the number of open file descriptors.
     """
-    if sys.platform.startswith(('linux', 'freebsd')):
+    if sys.platform.startswith(('linux', 'freebsd', 'darwin')):
         try:
-            names = os.listdir("/proc/self/fd")
-            # Substract one because listdir() opens internally a file
-            # descriptor to list the content of the /proc/self/fd/ directory.
+            fd_dir = '/dev/fd' if sys.platform == 'darwin' else '/proc/self/fd'
+            names = os.listdir(fd_dir)
+            # Subtract one because listdir() opens internally a file descriptor
+            # to list the contents of fd_dir.
             return len(names) - 1
         except OSError as exc:
             if exc.errno != errno.ENOENT:
-- 
2.31.1.498.g6c1eba8ee3d-goog

